### DISCUSSION TITLE
Skip delayed source w/ single track

### DISCUSSION THREAD

**Discussion by gammaw** [savonet/liquidsoap#2791] [User Support]
Originating from https://github.com/savonet/liquidsoap/issues/2766 I have an unwanted behaviour where a playlist with a single track in it will be faded out at the and of a scheduled time slot (track too long for the slot). The functionality we'd like to implement is that in such a case we skip the source so that the track starts from the beginning when the source is scheduled next time. 

The source is something like this: 

```ruby
sources = [ 
    ({!live_enabled}, live),    #Live DJ
    ({1w and 8h-8h58m}, delay(3600.,playlist_szedd_a_labad)),             # Monday
...
]

def fading_transition(a,b) =
   #the if statement is to skip certain sources only, delay_16 is just an example
    if source.id(a) == "delay_16" then
        log("Exiting show, skip source.")
        thread.run(delay=10.,{source.skip(a)})
    end
    sequence([
    fade.out(track_sensitive=false,duration=5.,a),
    playlist_jingle_station_id,    
    fade.in(track_sensitive=false,duration=5.,b)
    ])    
end
...
transitions = list.map(fun (_) -> fading_transition, sources)

radio = switch(id="schedule_switch", track_sensitive=false, transitions=transitions, sources)

```

What happens is: 
1. Track plays and gets faded out. 
2. The source (with a single track only) will be skipped. 
3. Next week, the source will be unavailable (probably because it hits the last track of the playlist). 
4. Next next week, the source will be available again (starting the track from the beginning). 

What is expected to happen*: 
1. Track plays and gets faded out. 
2. The source (with a single track only) will be skipped. 
3. Next week, the track will start from the beginning. 

The reason why we use `delay` is because we don't want the track to start over when it finishes _before_ the end of the scheduled time slot. 

*Does anybody have an idea how to implement the desired behavior? 

PS: we're running v2.1.2.

**Comment by toots** [comment]
Thanks for this report. I might have moved to quickly in converting it to a discussion.

We should understand why the playlist becomes unavailable. After the `skip` command is issued, the playlist should move to the next track. If this is the last track, the playlist should simply loop, at least if you use the normal playlist operator.

Do you have some logs around:
* The time the `skip` command is issued?
* The time the playlist is tried again?

**Comment by gammaw** [comment]
@toots good to know that my assumption about how a `skip` in a single-track source is supposed to work. I can imagine that the issue has to do with `delay`. Here's the log, with some additional print-outs of source IDs, e.g., the source `fail_3`, which may give you a clue. I manually added comments to mark the spots of scheduling and the corresponding example script. 

``` ruby
# Delayed source w/ one single track first scheduled at 15:19 in this example using specs ({7w and (15h19m-15h20m or 15h21m-15h22m)}, delay(30.,playlist_mmn_radio))
2022/12/11 15:19:00 [playlist_mmn_radio:3] Prepared "/var/azuracast/stations/lahmacun_radio/media/die_zahlen_von_0_bis_1000.mp3" (RID 71).
2022/12/11 15:19:00 [schedule_switch:3] Switch to delay_16 with transition.
2022/12/11 15:19:00 [lang:3] Fading transition starts
2022/12/11 15:19:00 [lang:3] a source id: cue_playlist_off_air_ambient
2022/12/11 15:19:00 [lang:3] b source id: delay_16
2022/12/11 15:19:00 [metadata.map_2:3] Inserting missing metadata.
[mp3float @ 0x7f26b4155440] Could not update timestamps for skipped samples.
2022/12/11 15:19:05 [cue_playlist_mmn_radio:3] Cueing in...
# Not sure what it is, I kept it here in case it has any relevance but I don't think so
Incorrect BOM value
Error reading lyrics, skipped
Incorrect BOM value
Error reading lyrics, skipped
# Delayed source is faded out at 15:20 (as per specs)
2022/12/11 15:20:00 [schedule_switch:3] Switch to cue_playlist_off_air_ambient with transition.
2022/12/11 15:20:00 [lang:3] Fading transition starts
2022/12/11 15:20:00 [lang:3] a source id: delay_16
2022/12/11 15:20:00 [lang:3] b source id: replay_metadata_0
2022/12/11 15:20:00 [lang:3] Exiting show, skip source.
2022/12/11 15:20:00 [metadata.map_2:3] Inserting missing metadata.
# Delayed source is skipped after 10s delay (see thread in script)
2022/12/11 15:20:10 [lang:3] Skip source: delay_16
# Delayed source is scheduled again at 15:21 (as per specs)
2022/12/11 15:21:00 [schedule_switch:3] Switch to delay_16 with transition.
2022/12/11 15:21:00 [lang:3] Fading transition starts
2022/12/11 15:21:00 [lang:3] a source id: cue_playlist_off_air_ambient
2022/12/11 15:21:00 [lang:3] b source id: replay_metadata_1
2022/12/11 15:21:00 [metadata.map_2:3] Inserting missing metadata.
2022/12/11 15:21:05 [playlist_mmn_radio:3] Prepared "/var/azuracast/stations/lahmacun_radio/media/die_zahlen_von_0_bis_1000.mp3" (RID 74).
# THIS IS THE ISSUE: Delayed source is available but failed, why??
2022/12/11 15:21:05 [schedule_switch:3] Switch to cue_playlist_off_air_ambient with forgetful transition.
2022/12/11 15:21:05 [lang:3] Fading transition starts
2022/12/11 15:21:05 [lang:3] a source id: fail_3
2022/12/11 15:21:05 [lang:3] b source id: replay_metadata_2
2022/12/11 15:21:05 [playlist_jingle_station_id:3] Prepared "/var/azuracast/stations/lahmacun_radio/media/50_cent.mp3" (RID 69).
[mp3float @ 0x7f26b4298c80] Could not update timestamps for skipped samples.
2022/12/11 15:21:05 [cue_playlist_jingle_station_id:3] Cueing in...
2022/12/11 15:21:05 [metadata.map_2:3] Inserting missing metadata.
# After 30s (as specified in the delay parameter - see specs) the delayed source becomes available again (still in the scheduled slot)
2022/12/11 15:21:35 [schedule_switch:3] Switch to delay_16 with transition.
2022/12/11 15:21:35 [lang:3] Fading transition starts
2022/12/11 15:21:35 [lang:3] a source id: cue_playlist_off_air_ambient
2022/12/11 15:21:35 [lang:3] b source id: replay_metadata_3
2022/12/11 15:21:35 [metadata.map_2:3] Inserting missing metadata.
[mp3float @ 0x7f26b428c6c0] Could not update timestamps for skipped samples.
2022/12/11 15:21:40 [cue_playlist_mmn_radio:3] Cueing in...
# Delayed source is faded out at 15:22 (as per specs)
2022/12/11 15:22:00 [schedule_switch:3] Switch to cue_playlist_off_air_ambient with transition.
2022/12/11 15:22:00 [lang:3] Fading transition starts
2022/12/11 15:22:00 [lang:3] a source id: delay_16
2022/12/11 15:22:00 [lang:3] b source id: replay_metadata_4
2022/12/11 15:22:00 [lang:3] Exiting show, skip source.
2022/12/11 15:22:00 [metadata.map_2:3] Inserting missing metadata.
2022/12/11 15:22:10 [lang:3] Skip source: delay_16
```

**Comment by toots** [comment]
Thanks for these details. As far as I understand, the use of `delay` seems to be creating issues. Are you expecting the delay to be canceled when skipping? It seems to me that the specs are respected in these logs, i.e. the source waits for the specified delay to be ready again.

We could definitely add a method to allow to skip _and_ reset delay so that the source is immediately available again, does that source correct?

**Comment by gammaw** [comment]
>Thanks for these details. As far as I understand, the use of delay seems to be creating issues. Are you expecting the delay to be canceled >when skipping? It seems to me that the specs are respected in these logs, i.e. the source waits for the specified delay to be ready again.

Not really, I expect the delayed source to be available after the skip is called _and_ the time in delay parameter passed. Here, it is not the case, only one "delay round" later. As I see it, the `skip` leaves a delayed single-track source in an "end of last available track" (EoLAT) state instead of the "start of the single-track" (SoT). When the source becomes available again, it will not play anything, it will only do the transiton from EoLAT -> SoT. Does it make sense? 

>We could definitely add a method to allow to skip and reset delay so that the source is immediately available again, does that source >correct?

Yes, something like that would be awesome! If there was a `source.reset` kind of command (which results in cueing to the start of a track in a playlist source), there would be no need of `skip`, right? 

I can also imagine that you change the semantics of skipping for a `delay(source)` type source so that the EoLAT state is not a valid state. I don't see a use case for it (becoming available in the EoLAT state), but if there may be one, a parameter could decide (e.g., `delay(source, skip_to_track_start=true)`).

**Reply by toots** [reply]
Thanks for the response, that makes a lot of sense now. I pushed some changed forcing the `delay` operator to reset to its start of next track state when calling skip, would you mind testing it: https://github.com/savonet/liquidsoap/actions/runs/3741832218

**Reply by toots** [reply]
Or maybe this one: https://github.com/savonet/liquidsoap/actions/runs/3741934004 :-)

**Reply by toots** [reply]
Nevermind I overlooked the fact that this was already done.. Give me a minute to re-read all that. Thanks!

**Reply by gammaw** [reply]
Hi @toots, not sure what's the status here. Are you planning to do a code update or is there a way to work it around with existing functionality? Thanks!

**Reply by toots** [reply]
I have let this one slip, gonna try to come back to it shortly.

**Comment by Non-Name-c** [comment]
123
